### DISCUSSION TITLE
Error "failed to get type: timeout" for oauth2 token calls

### DISCUSSION THREAD

**Discussion by anup-krai** [Kong/kong#10307] [Help]
We are using Kong with KIC with Postgres DB in Kubernetes environment. We have a route for oauth2 token generation `/oauth2/token` and [oauth2 plugin](https://docs.konghq.com/hub/kong-inc/oauth2/) is attached to this route.

Sometimes for this route we see below error messages in logs and Kong is returning HTTP 500 error code for this.

Error Log -
```
[error] 1097#0: *20435594 [kong] init.lua:297 [oauth2] /usr/local/share/lua/5.1/kong/plugins/oauth2/access.lua:694: [postgres] receive_message: failed to get type: timeout
```
And after this Kong responds with HTTP 500 with below message -
```
{"message":"An unexpected error occurred"}
```
Kong latency for this call is around 42 seconds. We are using the default value of [pg_timeout](https://github.com/Kong/kong/blob/master/kong.conf.default#L1134) variable.

What could be the reason for this error and this much high response time?

**Comment by ghost** [comment]
Looks related to this issue: #6137

Which version of Kong that you used?

**Reply by anup-krai** [reply]
Kong 2.7.1
KIC 2.2.1

**Reply by ghost** [reply]
I don't know much about this issue, and cc @fffonion 

I think you can try to verify if this problem exists on the latest version of Kong.

**Reply by fffonion** [reply]
@anup-krai Please try to debug as the steps for #6137, the timeout itself is most likely caused by a unreleased lock in postgres, or excessive high load on postgres.

**Reply by anup-krai** [reply]
Hi @fffonion , We don't see any lock in Postgres. From the Postgres logs I see only below two queries are taking more than a second -
```
LOG:  duration: 5908.564 ms  statement:   SELECT "id", EXTRACT(EPOCH FROM "created_at" AT TIME ZONE 'UTC') AS "created_at", "name", "al
gorithm", "hash_on", "hash_fallback", "hash_on_header", "hash_fallback_header", "hash_on_cookie", "hash_on_cookie_path", "slots", "healthchecks", "tags", "host_header", "client_certificate_id", "ws_id"
            FROM "upstreams"
        ORDER BY "id"
           LIMIT 1001; 
```
```
LOG:  duration: 3270.302 ms  statement:   SELECT "id", EXTRACT(EPOCH FROM "created_at" AT TIME ZONE 'UTC') AS "created_at", "name", "a
lgorithm", "hash_on", "hash_fallback", "hash_on_header", "hash_fallback_header", "hash_on_cookie", "hash_on_cookie_path", "slots", "healthchecks", "tags", "host_header", "client_certificate_id", "ws_id"
            FROM "upstreams"
        ORDER BY "id"
           LIMIT 1001;
```
Apart from these all the queries are running in ms. Also we see high Kong latencies. Anything else we can check on this?

**Reply by ghost** [reply]
check if the load on postgres is high when experiencing delays.

**Reply by anup-krai** [reply]
Postgres side load is normal. Only from Kong side we see when TPS increases the latency spikes up.

**Reply by anup-krai** [reply]
One more thing I noticed that we have [pg_timout](https://github.com/Kong/kong/blob/master/kong.conf.default#L1170) value set at 10 sec and Kong latency is going on till 40 sec for oauth2 token generation call. Why timeout is not happening at 10 sec.

**Reply by ghost** [reply]
> One more thing I noticed that we have [pg_timout](https://github.com/Kong/kong/blob/master/kong.conf.default#L1170) value set at 10 sec and Kong latency is going on till 40 sec for oauth2 token generation call. Why timeout is not happening at 10 sec.

check `pg_timeout` in `path/to/kong/.kong_env`

```
cat ./.kong_env | grep pg_timeout
```

**Reply by ghost** [reply]
and I noticed that the command line parameter `--db-timeout     (default 60)` used when starting Kong will override the `pg_timeout` setting in the `kong.conf`. Is this the expected behavior? cc @fffonion 

ref: 

1 https://github.com/Kong/kong/blob/d6e1d359e123f3423ce90af9863a0570eecef280/kong/cmd/init.lua#L69

here args would be like:

```json
{
  conf = "./kong.conf",
  db_timeout = 60,
  lock_timeout = 60,
  prefix = "/home/sabertobihwy/workspace/kong-dev/kong",
  v = false,
  vv = false
}
```

2. https://github.com/Kong/kong/blob/d6e1d359e123f3423ce90af9863a0570eecef280/kong/cmd/start.lua#L54

`conf.pg_timeout` would be `60000`, as the default `db_timeout` * 1000
